### PR TITLE
request methods return either a `Document` or a `NonHalResponse`

### PR DESCRIPTION
**Pull request by pezra**
Either works with `ExHal.url`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +29,8 @@ iex> doc = ExHal.client
 %ExHal.Document{...}
 ```
 
+### Navigation
+
 Now we have an entry point to the API we can follow links to navigate around.
 
 ```exlixir
```
Comment (alakra) [inline README.md]: minor typo here

File: lib/exhal.ex [hunk 2]
```diff
@@ -114,50 +115,13 @@ defmodule ExHal do
...
-  Returns `{:error, %ExHal.Error{...}}` if request is an error
-          `{:ok,    %ExHal.Document{...}}` if not
-  """
-  def post(a_doc, name, body) do
-    case figure_link(a_doc, name, false) do
-      {:error, e} -> {:error, e}
-      {:ok, link} -> Link.post(link, body, a_doc.client)
-    end
-  end
+  defdelegate [follow_link(a_doc, name),
```
Comment (alakra) [inline lib/exhal.ex]: Yay!

File: lib/exhal/navigation.ex [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                             {:ok, List.first(ls)}
+                           else
+                             {:error, %Error{reason: "multiple choices"}}
+                           end
+
+      [l] -> {:ok, l}
+    end
+  end
+
+  defp _follow_link(client, link, tmpl_vars, opts) do
```
Comment (alakra) [inline lib/exhal/navigation.ex]: I haven't seen this pattern before, what's the goal of this?

**Comment by alakra** [comment]
This looks good.

File: README.md [hunk 4]
```diff
@@ -68,6 +70,34 @@ iex> ExHal.follow_links(doc, "nonexistent")
...
+       { "href": "http://example.com/normal" }]
+    }
+  }
+  """
+
+# make a request that returns a HAL response
+iex> ExHal.put(doc, "self", name_change)
+{:ok, %ExHal.Document{...}}
+
+# make a request that just returns a response without without a body
```
Comment (lmarlow) [inline README.md]: Two `without without`s

File: README.md [hunk 5]
```diff
@@ -68,6 +70,34 @@ iex> ExHal.follow_links(doc, "nonexistent")
...
+iex> {:ok, resp} = ExHal.post(doc, "add-child", "{\"name\": \"child\"}")
+{:ok, %ExHal.NonHalResponse{status_code: 201, headers: ["Location" => "http://example.com/child", ...], body: ""}}
```
Comment (lmarlow) [inline README.md]: Should this really be `:ok`?

Comment (alakra) [inline README.md]: I think `:ok` is fine here, in this case it's not an error and the non-hal response is possible return value.  For context, our metric creation api takes a hal response but returns and empty body and a location header.

File: lib/exhal/navigation.ex (hunk 3 above)
Comment (pezra) [inline lib/exhal/navigation.ex]: i just wanted a "follow_link" function that is entirely different from the public `follow_link` function. The `_` prefix just uniqufies the name.

**Comment by pezra** [comment]
released as 4.0.0

File: lib/exhal/navigation.ex (hunk 3 above)
Comment (lmarlow) [inline lib/exhal/navigation.ex]: I often see erlang code with public API like `follow_link` and internal function `do_follow_link` is where the work is finally done.

**Pull request merged**